dj/construction/extract.py

```diff
@@ -0,0 +1,541 @@
+"""
+
+from dataclasses import dataclass, field
+from itertools import chain
+from typing import Dict, Generator, List, Optional, Set, Tuple, Union
+
+from sqlalchemy.orm.exc import NoResultFound
+from sqlmodel import Session, select
+
+from dj.errors import DJError, DJException, ErrorCode
+from dj.models.node import Node, NodeType
+from dj.sql.parsing.ast import (
+    Alias,
+    Column,
+    Named,
+    Query,
+    Select,
+    Table,
+    TableExpression,
+)
+from dj.sql.parsing.backends.sqloxide import parse
+
+
+def make_name(namespace, name="") -> str:
+    """utility taking a namespace and name to make a possible name of a DJ Node"""
+    ret = ""
+    if namespace:
+        ret += ".".join(name.name for name in namespace.names)
+    if name:
+        ret += ("." if ret else "") + name
+    return ret
+
+
+class CompoundBuildException:
+    """
+    Exception singleton to optionally build up exceptions or raise
+    """
+
+    errors: List[DJError]
+    _instance: Optional["CompoundBuildException"] = None
+    _raise: bool = True
+
+    def __new__(cls, *args, **kwargs):
+        if not cls._instance:
+            cls._instance = super(CompoundBuildException, cls).__new__(
+                cls, *args, **kwargs
+            )
+            cls.errors = []
+        return cls._instance
+
+    def reset(self):
+        """
+        Resets the singleton
+        """
+        self._raise = True
+        self.errors = []
+
+    def clear(self):
+        """
+        Erases stored errors
+        """
+        self.errors = []
+
+    def set_raise(self, raise_: bool):
+        """
+        Set whether to raise caught exceptions or accumulate them
+        """
+        self._raise = raise_
+
+    def catch(self, error: DJError, message: Optional[str]):
+        """
+        Accumulate DJ exceptions
+        """
+        if self._raise:
+            raise DJException(
+                message=message or error.message,
+                errors=[error],
+            )
+        self.errors.append(error)
+
+    def __str__(self) -> str:
+        plural = "s" if len(self.errors) > 1 else ""
+        error = f"Found {len(self.errors)} issue{plural}:\n"
+        return error + "\n\n".join(
+            "\t" + str(type(exc).__name__) + ": " + str(exc) + "\n" + "=" * 50
+            for exc in self.errors
+        )
+
+
+def get_dj_node(
+    session: Session,
+    node_name: str,
+    kinds: Optional[Set[NodeType]] = None,
+) -> Optional[Node]:
+    """Return the DJ Node with a given name from a set of node types"""
+    query = select(Node).filter(Node.name == node_name)
+    match = None
+    try:
+        match = session.exec(query).one()
+    except NoResultFound:
+        kind_msg = " or ".join(str(k) for k in kinds) if kinds else ""
+        CompoundBuildException().catch(
+            error=DJError(
+                code=ErrorCode.UNKNOWN_NODE,
+                message=f"No {kind_msg} node `{node_name}` exists.",
```

Comment:
Something I just noticed, this would be more readable if the "No" in the beginning was a "does not exist" at the end.
Currently: `No NodeType.DIMENSION or NodeType.SOURCE or NodeType.TRANSFORM node revenue exists.`
Proposed: `NodeType.DIMENSION or NodeType.SOURCE or NodeType.TRANSFORM node revenue does not exist.`

I keep missing the "No" and reading "node revenue exists".

Reply:
I had the same exact thought and then I forgot to comment.

Reply:
Changed to be 

```python
No node `revenue` exists of kind NodeType.DIMENSION or NodeType.SOURCE or NodeType.TRANSFORM.
```
good? Now you don't even have to read the whole message for the idea "no node revenue"

Reply:
Yes that's much clearer. I'm nitpicking now but having `does not exist` somewhere would make it a tiny bit clearer at first glance.

Reply:
Much better.